src/edu/csus/ecs/pc2/ui/MessageManager.java

```diff
@@ -0,0 +1,48 @@
+// Copyright (C) 1989-2023 PC2 Development Team: John Clevenger, Douglas Lane, Samir Ashoo, and Troy Boudreau.
+package edu.csus.ecs.pc2.ui;
+
+import java.util.Vector;
+
+/**
+ * Manage a set of message listeners.
+ * 
+ * @author Douglas A. Lane <pc2@ecs.csus.edu>
+ *
+ */
+public class MessageManager {
+
+    /**
+     * The list of message listeners.
+     */
+    private static Vector<IMessageRecordListener> messageListenerList = new Vector<IMessageRecordListener>();
+
+    /**
+     * Add a listener into listener list
+     */
+    public static void addMessageListener(IMessageRecordListener listener) {
+        messageListenerList.addElement(listener);
+    }
+    
+    /**
+     * Remove  a listener into listener list
+     */
+    public static void deleteMessageListener(IMessageRecordListener listener) {
+        messageListenerList.removeElement(listener);
+    }
+
+
+    /**
+     * Send message to all listeners.
+     * 
+     * @param record - message
+     */
+    public static void fireMessageListener(MessageRecord record) throws Exception {
+        for (int i = 0; i < messageListenerList.size(); i++) {
+            messageListenerList.elementAt(i).messageAdded(record);
```

Comment:
Since a `MessageRecord` already has a defined `MessageScope`, it seems like this method should only add messages to listeners interested in messages of the appropriate scope.

I don't consider this a reason to not approve the PR, but it's one of those things that seems pretty simple to fix now, while the new code is fresh in the mind... whereas the alternative will be the "Someday" that never comes...

Reply:
If there were a someday task here, I would agree.   There are a number of potential future "later" plans to improve the users'   ability to address both the minor and significant conditions, this change serves as a basis for a message system that will both cleanup  existing less than adequate/helpful error handling and provide for better insight/tasks for users to do to handle mis-configuration, security and other items.     

Adding MessageScope provides a way to add new messaging without accidentially sending/showing existing messages to the users at the "wrong" place in the GUI.

Reply:
`If there were a someday task here, I would agree. `  I have no idea what this means.  My point was that the current code already defines `MessageScope`, and it already has code which inserts messages in Message Listeners.  It seems to me it should only insert those messages if they are in the scope which the listener is interested in... and that this would be a simply thing to add "now" instead of waiting for "someday".

But again, I don't think this rises to the level of prohibiting Approval.  I just hate to see things coded "almost correctly" when it appears to be a very minor amount of effort to do it "correctly".